### ISSUE TITLE
Idea: support for embeddings

### ISSUE THREAD

**Issue by janekb04** [OpenNMT/CTranslate2#1183]
Being able to just emit text embeddings is useful, mostly for semantic search. I mean adding a Python and/or C++ API for using only the embedding stage of a model that returns the embedding vector.

**Comment by guillaumekln** [comment]
Do you mean running encoder-only models like BERT? There is another issue for that #1008.

**Comment by janekb04** [comment]
@guillaumekln Not really. I mean something similar to the OpenAI's embedding endpoint.

OpenAI has a completion endpoint, which models roughly `completion(prompt: str) -> str`. This is what CTranslate currently provides.

They also have an embedding endpoint. It models `embedding(text: str) -> list[float]`. The input is an arbitrary text (though it is usually natural language) and the output is a vector that is supposed to encode the semantic meaning of the text. The cool thing about this is that we can use those embeddings to make a database. We store a bunch of `(str, list[float])` pairs that are some knowledge base. To search through it, we first "embed" our query and then find the vectors in the database that are the closest to ours. This is supposed to return vectors whose texts have a similar meaning to our query, and hence, which are useful for answering it.

As far as I understand, every LLM computes these embeddings. Even a decoder-only model has to convert the input text string to its own vector representation of it. So, as far as I understand, this means only running the first layer(s) of an LLM and returning the hidden tensor representation of the input text. The idea behind this is that the LLM must be able to make good quality embeddings as it uses them itself, in later layers, to answer its queries.

**Comment by guillaumekln** [comment]
As far as I understand they trained models specifically for this task. Do you know if we can find equivalent models somewhere? Maybe it can also work with generic LMs as you mentioned.

> So, as far as I understand, this means only running the first layer(s) of an LLM and returning the hidden tensor representation of the input text.

The approach would need to be specified exactly. I don’t really know how and where they get the embeddings from the model.

**Comment by vancoykendall** [comment]
@janekb04 While the first layer of a transformer is called the embedding layer, the output of this layer isn't what OpenAI's embedding endpoint returns. They train an encoder transformer and use the last hidden vector representation of a special embedding token as the embedding vector.

Here is the description from OpenAI's [paper](https://arxiv.org/pdf/2201.10005.pdf) on their text embeddings model
> 2.1. Model
> Given a training pair (x, y), a Transformer (Vaswani et al., 2017) encoder E is used to process x and y independently. The encoder maps the input to a dense vector representation or embedding (Figure 2). We insert two special token delimiters, [SOS] and [EOS], to the start and end of the input sequence respectively. The hidden state from the last layer corresponding to the special token [EOS] is considered as the embedding of the input sequence.

So I believe this would be supported by the issue #1008 that @guillaumekln mentioned.

**Comment by janekb04** [comment]
Thanks for the explanation. I have indeed assumed that OpenAI's embeddings are the input token embeddings given to the transformer. But it makes more sense that they are actually the outputs form the encoder. Its interesting to hear that as I thought that GPT-3 is a decoder-only model.

Then, what is the difference between an encoder-only model and a decoder-only model? I did some reading on transformers. As far as I understand, the difference between the encoder and the decoder is that the decoder has two self-attention layers. The first uses masked self-attention to avoid attending to future outputs. The second one is encoder-decoder attention that takes the value and key matrices from the encoder and queries from masked self-attention. Given that decoder-only models are already supported, I'm not sure what would really be needed to support encoder-only models as well. To me it seems like an encoder-only model is a decoder-only model, but without the second self-attention layer and masking applied. Alternatively, why can't the encoder just be ripped out of a transformer and run on its own?

![image](https://user-images.githubusercontent.com/13035076/234395383-5235083e-e08e-4e65-946b-2c2bfe6747b8.png)

**Comment by vancoykendall** [comment]
In this context I think of there being three types of transformer architectures: encoders, decoders with cross-attention, and decoders without cross-attention (by cross-attention I mean the encoder-decoder attention you are referring to). 

There is only one type of encoder since it will have the same architecture when used on its own (encoder-only model) and when used with a decoder (encoder-decoder models). A decoder's architecture, on the other hand, will differ depending on if it's used on its own (decoder-only) and if it's used with an encoder (encoder-decoder models). When a decoder is used on its own as a decoder-only model, *it doesn't have the cross-attention layer* (since there are no encoder embeddings to perform cross-attention on). This means the decoder-only model architecture is identical to the encoder model architecture, except for the causal attention mask applied in the decoder's attention layer.

I haven't actually looked at the c++ code for this library, but likely the `ctranslate2.Generator` class uses a decoder-only model and `ctranslate2.Translator` uses both an encoder and decoder model (where the decoder has the cross-attention layer). So most of the code to support encoder models likely already exists, with the exception of a class to expose the embeddings.

> Alternatively, why can't the encoder just be ripped out of a transformer and run on its own?

An encoder can be ripped out of an encoder-decoder model and run on its own. However, it probably wouldn't be that useful unless it was either further trained without the decoder or at least fine-tuned on some task.

**Comment by janekb04** [comment]
Thanks for the comment. I understand now that the main difference between an encoder-only model and a decoder-only model is in how they produce the output sequence. The encoder produces the entire output at once, while a decoder is autoregressive and produces it token by token. So the base architecture is the same, but the decoder applies masking to avoid looking ahead.

And an encoder taken out of an existing transformer might not yield satisfactory results as that is not the task it was trained for.

**Comment by guillaumekln** [comment]
Thanks for the discussion!

So it seems that I can close this issue in favor of https://github.com/OpenNMT/CTranslate2/issues/1008.

**Comment by janekb04** [comment]
@vancoykendall Are these definitions of how the three types of transformers work certainly correct?

I recently came across [a post](https://www.linkedin.com/posts/yann-lecun_a-survey-of-llms-with-a-practical-guide-and-activity-7057527966540386304-M4_2?utm_source=share&utm_medium=member_ios) by [Yann LeCun](https://www.linkedin.com/in/yann-lecun) where he writes:

> The architecture nomenclature for LLMs is somewhat confusing and unfortunate.
> What's called "encoder only" actually has an encoder and a decoder (just not an auto-regressive decoder).
> What's called "encoder-decoder" really means "encoder with auto-regressive decoder"
> What's called "decoder only" really means "auto-regressive encoder-decoder"

So, doesn't this mean that every transformer has both an encoder and a decoder, with the encoder-decoder cross-attention layer always included? Or am I misunderstanding something?

**Comment by guillaumekln** [comment]
> So, doesn't this mean that every transformer has both an encoder and a decoder, with the encoder-decoder cross-attention layer always included?

No. I think LeCun is trying to be pedantic here but it just makes things more confusing.

Your previous comment is correct:

> The encoder produces the entire output at once, while a decoder is autoregressive and produces it token by token. So the base architecture is the same, but the decoder applies masking to avoid looking ahead.

We also say that the encoder self-attention is bidirectional, while the decoder self-attention is unidirectional (via masking).
